android/src/main/java/com/tailscale/ipn/IPNService.kt

```diff
@@ -114,12 +115,24 @@ open class IPNService : VpnService(), libtailscale.IPNService {
     }
     b.setUnderlyingNetworks(null) // Use all available networks.
 
-    // Prevent certain apps from getting their traffic + DNS routed via Tailscale:
-    // - any app that the user manually disallowed in the GUI
-    // - any app that we disallowed via hard-coding
-    for (disallowedPackageName in UninitializedApp.get().disallowedPackageNames()) {
-      Log.d(TAG, "Disallowing app: $disallowedPackageName")
-      disallowApp(b, disallowedPackageName)
+    val includedPackages: List<String> =
+        MDMSettings.includedPackages.flow.value?.split(",") ?: emptyList()
+    if (includedPackages.isNotEmpty()) {
+      // If an admin defined a list of packages that are exclusively allowed to be used via
+      // Tailscale,
+      // then only allow those apps.
+      for (packageName in includedPackages) {
+        Log.d(TAG, "Including app: $packageName")
+        b.addAllowedApplication(packageName)
+      }
+    } else {
+      // Otherwise, prevent certain apps from getting their traffic + DNS routed via Tailscale:
+      // - any app that the user manually disallowed in the GUI
+      // - any app that we disallowed via hard-coding
+      for (disallowedPackageName in UninitializedApp.get().disallowedPackageNames()) {
+        Log.d(TAG, "Disallowing app: $disallowedPackageName")
+        disallowApp(b, disallowedPackageName)
+      }
```

Comment:
If both lists are non-empty, log an error that you're ignoring the excluded list.

Reply:
disallowedPackageNames is never going to be empty. We have hardcoded exclusions.